Copyright Headers Tool

Cdok opened the pull request:
## Changes
- This PR adds a tool to either check for existing copyright headers, or add copyright headers if they are missing
- This tool is language agnostic (see note below), and can be pointed to any directory using a glob pattern


**Note:** this tool reads from the `headers/copyright.extension` files to select what header to insert, where `extension` is the target language

Currently only `.py`, `.jsx`, and `.js` have been added for testing.

## How To Test

- Clone the repo and checkout this branch (`copyright-headers`)
- `cd` into the `copyright` directory, and run `node copyright.js 'PATTERN'` where `PATTERN` is a glob pattern
- For example, if you cloned this repo in the same directory as the `progressive-web-sdk`, and needed to add copyright headers to all `.js` files in the `src` directory of the SDK, you would write:

```bash
node copyright.js '../../progressive-web-sdk/src/**/*.js' --fix
```

Note the optional `--fix` flag that can be passed in, which will add the headers to files which do not contain it.

Another flag `--update` will check the target files for outdated headers, and update them to the current year.

The default mode (no flags) will exit the process with an error if any of the targeted files are missing copyright headers.


### To test in dev mode (with chrome inspect)
`node --inspect --debug-brk copyright.js ../../progressive-web-sdk/bin/**/*.js`


## TODOs:
- [x] +1
- [x] Updated README
- [x] Updated CHANGELOG


## Resources:
[node-glob](https://github.com/isaacs/node-glob)

jeremywiebe commented on README.md:
```diff
@@ -104,3 +104,10 @@ instructions on using `pep8`, `pylint` and `pyflakes` to check Python code.
+
+## Copyright Headers
+
+All Mobify source files should contain copyright headers at the beginning of the file.
```
@johnboxall "*All*" or just source we intend to ship outside our control (ie. to the browser/device). I wouldn't anticipate server code needs these?

jeremywiebe commented on copyright/README.md:
```diff
@@ -0,0 +1,30 @@
+
+## Quick Start
+
+- Clone this repo
```
I've seen these formatted as a short terminal session... sometimes that's clearer.  Eg.

```sh
git clone git@github.com:mobify/mobify-code-style.git
cd mobify-code-style/copyright
node copyright.js ../**/*.js ../**/*.py
```

Cdok replied:
@jeremywiebe so ideally would we want something like 
```bash
npm install mobify-code-style
npm run lint:copyright
```

Or something along these lines?

I made the change in the README right now - so I should also add something like

```bash
"lint:copyright": "node_modules/mobify-code-style/copyright/copyright.js PATH --lint"
```

to your `package.json` - or is there an easier way than digging into the node modules in the npm script 😛 🤔

jeremywiebe commented on copyright/headers/copyright.js.txt:
```diff
@@ -0,0 +1,3 @@
+/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
+/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
```
Can we make the year dynamic?  It'd be a pain to have to update this every year. :)  When we run the tool it asserts the copyright as of the current year, I think that'd be enough.

jeremywiebe commented on package.json:
```diff
@@ -19,6 +19,7 @@
     "test": "cd docs/test; bash test.sh"
   },
   "dependencies": {
+    "glob": "^7.1.1",
```
Can we add this tool to the `"bin"` key just a few lines above?  That will make it much easier to use in projects that import this npm package.

jeremywiebe commented on copyright/copyright.js:
```diff
@@ -0,0 +1,106 @@
+                fs.readFile(item, (err, data) => {
+                    if (!this.hasCopyrightHeader(data)) {
+                        if (this.lintMode === true) {
+                            console.log(`\x1b[33m Missing copyright headers in \x1b[36m ${item}`)
```
It'd be good to use constants for these escape codes.  Without running the tool I have no idea what `\x1b[33m` does.  :)

jeremywiebe commented on copyright/copyright.js:
```diff
@@ -0,0 +1,106 @@
+/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
+/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
```
:)  

![imgres](https://cloud.githubusercontent.com/assets/77138/25680300/2c7ce42c-3005-11e7-95b8-cde1235b04f5.jpg)

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,106 @@
+            return true
+        }
+        return false
+    },
```
This function can just be `return data.indexOf(COPYRIGHT_STRING) >= 0`.

It doesn't actually use `this`, so it could even move outside this object and become even simpler.

jeremywiebe commented on copyright/copyright.js:
```diff
@@ -0,0 +1,106 @@
+const copyright = {
+    items: [],
+    missingHeaders: [],
+    copyrightText: '',
```
This assumes that the glob(s) you pass will only ever be of the same type. It might be better to keep a dict of extension -> header text so that you can evaluate on a per-file basis which header you need. 

Given that we'll probably only have a handful of file types supported, you could read those in before you start and have them handy throughout the run.

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,106 @@
+
+        const processedGlobs = args.map((dir) => {
+            return new Promise((resolve) => {
+                this.getHeaderText(dir)
```
If the `dir`s have different header text, this will cause them to step on each other.

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,106 @@
+                this.getHeaderText(dir)
+                glob(`${dir}`, (err, files) => {
+                    this.items.push(files[0])
+                    resolve()
```
Rather than using the `this.items` array, we can just pass `files[0]` to `resolve`. Then the `Promise.all` call below will resolve to an array of the same contents.

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,106 @@
+        } else {
+            this.copyrightText = fs.readFileSync(path)
+        }
+    },
```
With the files that are included here, we'll get an error for `.jsx` files, which are very common in the Progressive code.

jeremywiebe commented on copyright/copyright.js:
```diff
@@ -0,0 +1,106 @@
+const fs = require('fs')
+
+const copyright = {
+    items: [],
```
What are these items? :)  I think they're individual file paths, so renaming them to such would make this variable alot easier to "digest".  Maybe `filePaths: []` or even `filesToProcess: []`?

Cdok commented:
@DanielBrookRoberge @jeremywiebe thank you 😄  addressed the issues with namespace pollution, and refactored into more of a pipeline.

Ready for more feedback 🎉

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,109 @@
+            filenames.forEach((file) => {
+                const extension = file.match(/\.[0-9a-z]+/i)[0]
+                this.langs[extension] = file
+            })
```
The header files don't really need to have a `.txt` extension; if we leave it off we can use the same extension extraction logic for both this and the actual source files.

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,109 @@
+        const processedGlobs = args.map((folder) => {
+            return new Promise((resolve) => {
+                glob(`${folder}`, (err, file) => {
+                    resolve(file[0])
```
This looks like it only gets the first filename matching the glob, is that what you want?

Cdok replied:
Each matched `file` that's inside `glob()` callback is an Array of length 1
![screen shot 2017-05-04 at 10 51 27 am](https://cloud.githubusercontent.com/assets/16109837/25717597/cceaab8a-30b7-11e7-8d81-e303be2620ed.png)

From `node-glob`:

```javascript
var glob = require("glob")

// options is optional
glob("**/*.js", options, function (er, files) {
  // files is an array of filenames.
  // If the `nonull` option is set, and nothing
  // was found, then files is ["**/*.js"]
  // er is an error object or null.
})
```

So it seems like this should be an array of all of the matches - but instead the functionality is that the callback is fired 1 time for each matched file - and `file` is an array of length 1 🤔

DanielBrookRoberge replied:
So then this wouldn't work with the `resolve` method since it would only grab the first result (subsequent calls to `resolve` do nothing). 

It might just be simpler to use `glob.globSync` and then flatten the resulting array of arrays.

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,109 @@
+        })
+
+        Promise.all(processedGlobs).then((files) => {
+            const filesMissingHeader = files.filter((file) => {
```
This variable appears to contain filenames that do contain the header, rather than ones that are missing the header, right?

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,109 @@
+                    }
+                } else {
+                    return true
+                }
```
This could be simpler and less nested if you start out with 
```
if (hasCopyrightHeader) {
    return true
}
```

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,109 @@
+            const filesMissingHeader = files.filter((file) => {
+                const content = fs.readFileSync(file)
+                const copyrightStr = 'Copyright (c)'
+                const hasCopyrightHeader = content.indexOf(copyrightStr) >= 0
```
In Node, we can use `String.prototype.includes` for something like this.

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,108 @@
+                        console.log(`${green}Copyright header succesfully written into ${magenta}${file}`)
+                        return true
+                    }
+                }
```
Having restructured things in this way, we don't actually need the `else` block; the `return` statement will exit the function.

jeremywiebe commented on package.json:
```diff
@@ -12,7 +12,8 @@
   "bin": {
-    "lint-md": "docs/bin/lint-md"
+    "lint-md": "docs/bin/lint-md",
+    "glob": "^7.1.1"
```
Hrm... is this in the right place?  Isn't this a dependency?

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,106 @@
+
+        const processedGlobs = args.map((folder) => {
+            files.push(glob.sync(folder))
+        })
```
If you return `glob.sync(folder)` from the map function, then you won't need the `files` array; `processedGlobs` will contain what you want.

DanielBrookRoberge commented on copyright/copyright.js:
```diff
@@ -0,0 +1,106 @@
+                    return true
+                }
+            }
+        })
```
Thinking about this some more, we don't really need this array, just a flag for whether an error has occurred. Then this can be a `forEach` that just sets the flag rather than worrying about return values.

Cdok commented:
Just added the node environment shebang, and updated the readme to reflect the new installation instructions.

On the thought of shebangs - should we consider accommodating for them in this tool? I'm curious as to whether or not any issues would arise out of inserting a copyright header before a shebang (whether it be issues from just our linter or otherwise) @jeremywiebe  thoughts? 🤔

Cdok commented:
Tests via `npm install` confirmed to be working on the SDK, noting that the glob needs to be wrapped in quotes to hit all of the nested directories.

Updated the readme and `--help` flag to reflect this

jeremywiebe commented on copyright/README.md:
````diff
@@ -0,0 +1,45 @@
+Then run
+
+```bash
+npm run copyright:lint
````
This should be `copyright:fix` right?

jeremywiebe commented on copyright/README.md:
```diff
@@ -0,0 +1,45 @@
+
+## Lint Mode
+
+Passing the `--fix` flag to this tool will disable lint mode.
```
I think this is incorrect now also.  You pass nothing to lint and pass `--fix` to modify files.

jeremywiebe commented:
@Cdok I didn't think of shebangs... yes, we probably should account for them.

Cdok commented on copyright/copyright.js:
```diff
@@ -0,0 +1,128 @@
+                // accomodate for shebang and insert before header
+                if (contentStr[0].indexOf('#!') >= 0) {
+                    shebang = contentStr.shift()
+                    content = contentStr.join('\n')
```
I'm not sure if this check for the shebang _feels_ right. It's currently working, but the string manipulation here feels a bit awkward.

<img width="705" alt="screen shot 2017-05-04 at 4 15 12 pm" src="https://cloud.githubusercontent.com/assets/16109837/25728577/e3c39754-30e4-11e7-8c6a-cccc1cd26b5f.png">

Cdok commented:
@jeremywiebe oneeeee more thing that I hadn't thought of 😏  

I just added an optional `--update` flag so that in 2018 the headers can be updated by passing the flag, or we could just use
```json
"copyright:lint": "copyright './src/**/*.js' --update",
```
in the `npm run` script, and let the updates happen automagically.

Either way, the option is there for both right now. 

Lastly, the shebang check functionality is up in this PR now too.

I've tested it against `bin` in the SDK.

With that said, I _think_ this PR is ready for the last feedback pass 👀
